### ISSUE TITLE
[BUG] TensorFlow backprop gradient non-zero imaginary part

### ISSUE THREAD

**Issue by antalszava** [PennyLaneAI/pennylane#2872]
### Expected behavior

Computing the gradient of circuits using backprop with TF yields real-valued (or complex with negligible imaginary part) outputs.

### Actual behavior

Some results have non-negligible imaginary parts.

### Additional information

The gate parameters with the TF interface are being cast to a complex type in the Operation definitions. See `if qml.math.get_interface(theta) == "tensorflow"` conditions and the PR introducing this logic.

### Source code

```shell
import pennylane as qml
import tensorflow as tf

dev = qml.device("default.qubit", wires=2)

phi = qml.numpy.array(0.1)
psi_0 = tf.Variable(0.1, dtype=tf.complex128)
psi_1 = tf.Variable(0.2, dtype=tf.complex128)
psi_2 = tf.Variable(0.3, dtype=tf.complex128)
psi_3 = tf.Variable(0.4, dtype=tf.complex128)


phi = qml.math.cast_like(phi, 1j)

init_state = tf.Variable([psi_0, psi_1, psi_2, psi_3], dtype=tf.complex128)
norm = tf.norm(init_state)
init_state = init_state / norm

@qml.qnode(dev, interface="tf", diff_method="backprop")
def circuit(phi):
    qml.QubitStateVector(init_state, wires=[0, 1])
    qml.RZ(phi, wires=[0])
    return qml.expval(qml.PauliY(0))

phi = tf.Variable(phi, dtype=tf.complex128)

with tf.GradientTape() as tape:
    result = circuit(phi)

res = tape.gradient(result, phi)
print(res)

# Result: tf.Tensor((0.7296697212038854+0.07321117220767401j), shape=(), dtype=complex128)
```

### System information

```shell
PL dev
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

**Comment by dwierichs** [comment]
This is not too surprising I think, because the differentiation w.r.t complex parameters is not uniquely defined.
When using `tf.float64` for `phi` instead, everything is as expected:
```python
phi = tf.Variable(phi, dtype=tf.float64)

with tf.GradientTape() as tape:
    result = circuit(phi)

res = tape.gradient(result, phi)
```
```pycon
>>> print(res)
tf.Tensor(0.7296697212038854, shape=(), dtype=float64)
```

I tried to dig up at which point exactly the difference between the holomorphic and componentwise differentiation kicks in, but I didn't find it right away. Let me know in case this is an important issue/serious bug somewhere :)

**Comment by dwierichs** [comment]
I do not think this is a bug and would vote for closing. @josh146 What do you think?

**Comment by josh146** [comment]
yep this sounds more TF related!
